### PR TITLE
Introduce Context Menu Commands (v5): User Commands / Message Commands

### PR DESCRIPTION
**Pull request by aasmart**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [X] I have checked the PRs for upcoming features/bug fixes.
- [X] I have read the [contributing guidelines][contributing].

### Changes

- [X] Internal code
- [X] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

This PR introduces context menus. This is identical to #1769, but for v5 instead of development

### How to Create a Context Menu Command
Creating a context menu command is as simple as creating a normal slash command. Instead, using the constructor CommandData(type, name) allows you to create a command of the type CommandType#USER_COMMAND or CommandType#MESSAGE_COMMAND with a name. Attempting to add choices, subcommands, descriptions, etc. does not work on context menu commands.

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/JDA.java [hunk 1]
```diff
@@ -590,6 +591,38 @@ default CommandCreateAction upsertCommand(@Nonnull String name, @Nonnull String
...
+     *         If null is provided or the name/description do not meet the requirements.
+     *         Also, if the command being created is of type {@link CommandType#SLASH}
+     *
+     * @return {@link CommandCreateAction}
+     *
+     * @see net.dv8tion.jda.api.entities.Guild#upsertCommand(String, String)
+     */
+    @Nonnull
+    @CheckReturnValue
+    default CommandCreateAction upsertCommand(CommandType commandType, @Nonnull String name)
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/JDA.java:620]:
```suggestion
    default CommandCreateAction upsertCommand(@Nonnull CommandType commandType, @Nonnull String name)
```

File: src/main/java/net/dv8tion/jda/api/JDA.java [hunk 2]
```diff
@@ -590,6 +591,38 @@ default CommandCreateAction upsertCommand(@Nonnull String name, @Nonnull String
...
+    {
+        Checks.check(commandType != CommandType.SLASH, "This command may not be of type slash command");
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/JDA.java:622]: Why?

File: src/main/java/net/dv8tion/jda/api/events/interaction/commands/MessageContextCommandEvent.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import javax.annotation.Nonnull;
+
+/**
+ * Indicates that a message command was used in a {@link MessageChannel}.
+ *
+ * <h2>Requirements</h2>
+ * To receive these events, you must unset the <b>Interactions Endpoint URL</b> in your application dashboard.
+ * You can simply remove the URL for this endpoint in your settings at the <a href="https://discord.com/developers/applications" target="_blank">Discord Developers Portal</a>.
+ */
+public class MessageContextCommandEvent extends GenericCommandEvent implements MessageCommandInteraction
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/events/interaction/commands/MessageContextCommandEvent.java:19]: Would it make more sense to just combine all context menu interactions into a single type that has individual (nonnull) getters:

- getTargetType / getCommandType
- getTargetMessage
- getTargetUser
- getTarget<...>

@DV8FromTheWorld what do you think? I think that would offer the same feature but require less work for extensions.

File: src/main/java/net/dv8tion/jda/api/events/interaction/commands/UserContextCommandEvent.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package net.dv8tion.jda.api.events.interaction.commands;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/events/interaction/commands/UserContextCommandEvent.java:1]: Copyright header

File: src/main/java/net/dv8tion/jda/api/events/interaction/commands/MessageContextCommandEvent.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
+package net.dv8tion.jda.api.events.interaction.commands;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/events/interaction/commands/MessageContextCommandEvent.java:1]: Copyright header

File: src/main/java/net/dv8tion/jda/api/events/interaction/commands/GenericCommandEvent.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
+package net.dv8tion.jda.api.events.interaction.commands;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/events/interaction/commands/GenericCommandEvent.java:1]: Copyright header

File: src/main/java/net/dv8tion/jda/api/interactions/commands/CommandType.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
+package net.dv8tion.jda.api.interactions.commands;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/CommandType.java:1]: Copyright header

File: src/main/java/net/dv8tion/jda/api/interactions/commands/MessageCommand.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
+package net.dv8tion.jda.api.interactions.commands;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/MessageCommand.java:1]: Copyright header

File: src/main/java/net/dv8tion/jda/api/interactions/commands/UserCommand.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
+package net.dv8tion.jda.api.interactions.commands;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/UserCommand.java:1]: Copyright header

File: src/main/java/net/dv8tion/jda/api/interactions/commands/build/BaseCommand.java [hunk 10]
```diff
@@ -73,14 +96,16 @@ public T setName(@Nonnull String name)
...
+     *         If the name is null or not between 1-100 characters.
+     *         Also, if the command is not of type {@link CommandType#SLASH}
      *
      * @return The builder, for chaining
      */
     @Nonnull
     @SuppressWarnings("unchecked")
     public T setDescription(@Nonnull String description)
     {
+        Checks.check(commandType == CommandType.SLASH, "You cannot modify this command's description");
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/build/BaseCommand.java:108]: Make this error more descriptive

File: src/main/java/net/dv8tion/jda/api/interactions/commands/interactions/CommandInteraction.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
+package net.dv8tion.jda.api.interactions.commands.interactions;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/interactions/CommandInteraction.java:1]: Copyright header

File: src/main/java/net/dv8tion/jda/api/interactions/commands/interactions/ContextMenuInteraction.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
+package net.dv8tion.jda.api.interactions.commands.interactions;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/interactions/ContextMenuInteraction.java:1]: Copyright header

File: src/main/java/net/dv8tion/jda/api/interactions/commands/interactions/UserCommandInteraction.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
+package net.dv8tion.jda.api.interactions.commands.interactions;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/interactions/commands/interactions/UserCommandInteraction.java:1]: Copyright header

**Review by MinnDevelopment — COMMENTED** [review]
This is just my first pass over it, there is definitely more work to be done.

**Comment by altrisi** [comment]
Are there any updates on this?

**Comment by oliver276** [comment]
A rework of interaction handling is next on the list of v5 tasks.  After this is done, context menus will be implemented.

I don't think this branch will be used, as it relies on the old (current) interactions.

**Comment by MinnDevelopment** [comment]
Superseded by #1971

**Pull request closed without merging**